File: 0028. Find the Index of the First Occurrence in a String.py
# Easy
# O(m + n)
# https://leetcode.cn/problems/find-the-index-of-the-first-occurrence-in-a-string/

class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        return haystack.find(needle)

# 這題用Python最快，直接調用find，如果是Java需要看一下KMP

# Java代碼
# class Solution {
#     public int strStr(String haystack, String needle) {
#         if (needle.length() == 0) return 0;
#         int[] next = new int[needle.length()];
#         getNext(next, needle);

#         int j = 0;
#         for (int i = 0; i < haystack.length(); i++) {
#             while (j > 0 && needle.charAt(j) != haystack.charAt(i)) 
#                 j = next[j - 1];
#             if (needle.charAt(j) == haystack.charAt(i)) 
#                 j++;
#             if (j == needle.length()) 
#                 return i - needle.length() + 1;
#         }
#         return -1;

#     }
    
#     private void getNext(int[] next, String s) {
#         int j = 0;
#         next[0] = 0;
#         for (int i = 1; i < s.length(); i++) {
#             while (j > 0 && s.charAt(j) != s.charAt(i)) 
#                 j = next[j - 1];
#             if (s.charAt(j) == s.charAt(i)) 
#                 j++;
#             next[i] = j; 
#         }
#     }
# }

File: 0001-0500/0028. Find the Index of the First Occurrence in a String.java
// Easy
// String
// Time:O(m*n) for brute force, Space:O(1)
// KMP -> O(m+n)
// https://leetcode.cn/problems/find-the-index-of-the-first-occurrence-in-a-string/

class Solution {
    // Iterate through haystack, start from each possible pos i
    // For each pos, compare chars in haystack and needle
    // If all chars in needle match, return the cur index
    // If no match is found after iteration, return -1
    public int strStr(String haystack, String needle) {
        int m = haystack.length(), n = needle.length();
        for (int i = 0; i <= m - n; i++) {
            int j = 0; // j use to iterate the needle
            // check if haystack[i + j] equal to needle[j]
            while (j < n && haystack.charAt(i + j) == needle.charAt(j)) {
                j++;
            }
            // if j == n, means we match all chars in needle
            if (j == n) return i;
        }
        return -1;
    }
}
